services/motion/builtin/testutils.go

```diff
@@ -0,0 +1,542 @@
+	deps[fakeBase.Name()] = fakeBase
+	deps[sens.Name()] = sens
+
+	return deps
+}
+
+func createWheeledBase(
+	t *testing.T,
+	ctx context.Context,
+	logger logging.Logger,
+	motors map[string]motor.Motor,
+	geomSize float64,
+) base.Base {
+	t.Helper()
+
+	reg, ok := resource.LookupRegistration(base.API, wheeled.Model)
+	test.That(t, ok, test.ShouldBeTrue)
+
+	wheeledConf := &wheeled.Config{
+		WidthMM:              200,
+		WheelCircumferenceMM: 2000,
+		Left:                 []string{leftMotorName},
+		Right:                []string{rightMotorName},
+	}
+
+	baseConf := resource.Config{
+		Name:                baseName,
+		API:                 base.API,
+		Frame:               &referenceframe.LinkConfig{Geometry: &spatialmath.GeometryConfig{R: geomSize}},
+		ConvertedAttributes: wheeledConf,
+	}
+	deps := resource.Dependencies{}
+	deps[motors[leftMotorName].Name()] = motors[leftMotorName]
+	deps[motors[rightMotorName].Name()] = motors[rightMotorName]
+
+	wBase, err := reg.Constructor(ctx, deps, baseConf, logger)
+	test.That(t, err, test.ShouldBeNil)
+	return wBase.(base.Base)
+}
+
+func createEncoders(t *testing.T, ctx context.Context, logger logging.Logger) map[string]encoder.Encoder {
+	t.Helper()
+	leftconf := resource.Config{
+		Name:                leftMotorName + "_encoder",
+		API:                 encoder.API,
+		ConvertedAttributes: &fakeencoder.Config{UpdateRate: int64(updateRate)},
+	}
+	rightconf := resource.Config{
+		Name:                rightMotorName + "_encoder",
+		API:                 encoder.API,
+		ConvertedAttributes: &fakeencoder.Config{UpdateRate: int64(updateRate)},
+	}
+	left, err := fakeencoder.NewEncoder(ctx, leftconf, logger)
+	test.That(t, err, test.ShouldBeNil)
+	right, err := fakeencoder.NewEncoder(ctx, rightconf, logger)
+	test.That(t, err, test.ShouldBeNil)
+	return map[string]encoder.Encoder{leftMotorName: left, rightMotorName: right}
+}
+
+func createMotors(
+	t *testing.T,
+	ctx context.Context,
+	logger logging.Logger,
+	encoders map[string]encoder.Encoder,
+) map[string]motor.Motor {
+	t.Helper()
+	motorMap := map[string]motor.Motor{}
+	for name, encoder := range encoders {
+		conf := resource.Config{
+			Name:                name,
+			API:                 motor.API,
+			ConvertedAttributes: &fakemotor.Config{Encoder: name + "_encoder", TicksPerRotation: 4000},
+		}
+
+		thisMotor, err := fakemotor.NewMotor(
+			ctx,
+			resource.Dependencies{encoder.Name(): encoder},
+			conf,
+			logger,
+		)
+		test.That(t, err, test.ShouldBeNil)
+		motorMap[name] = thisMotor
+	}
+	return motorMap
+}
+
+type noisyMovementSensor struct {
+	movementsensor.MovementSensor
+	mu         sync.Mutex
+	noise      spatialmath.Pose
+	queryCount int // Apply noise every other iteration when this is > 1
+}
+
+func (m *noisyMovementSensor) Position(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.queryCount++
+	if m.queryCount <= 10 || m.queryCount%2 == 0 {
```

Comment:
This is strange and I do not get why this exists.

Reply:
This is the part that introduces the "noisy" part of the `noisyMovementSensor`. As the underlying movement sensor is relying on other wrapped components to make the base actually traverse the planned path, it is less easy to inject noise than it was.

If we want tests to replan more than once, we cannot have a single offset, as the first replan will correct for that noise and a second replan will not occur. Thus we must oscillate whether or not we return a pose with the noise applied.